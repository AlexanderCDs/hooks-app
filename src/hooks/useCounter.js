/** 
 * @author: alexandercddev
 * @description: 
 * @date: 30/Septiembre/2021
**/
import { useState } from 'react';

export const useCounter = (inicialState = 10) => {
    const [counter, setCounter] = useState(inicialState);
    const increment = () => setCounter( counter + 1);
    const decrement = () => setCounter( counter - 1);
    const reset = () => setCounter(inicialState);

    return {
        counter,
        increment,
        decrement,
        reset
    }
}