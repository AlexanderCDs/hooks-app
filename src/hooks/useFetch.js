/** 
 * @author: alexandercddev
 * @description: 
 * @date: 30/Septiembre/2021
**/   
import { useState, useEffect, useRef } from "react";

export const useFetch = (url) => {
    const isMounted = useRef(true);
    const [state, setState] = useState({
        data: null,
        loading: true,
        error: null
    });

    useEffect(() => {
        return () =>{
            isMounted.current = false;
        }
    }, [])

    useEffect(() => {
        setState({
            data: null,
            loading: true,
            error: null
        });
        fetch( url )
            .then(response => response.json() )
            .then ( data => {
                if( isMounted.current ){
                    setState({
                        data,
                        loading: false,
                        error: null
                    });
                } 
            })
            .catch( () => {
                setState({
                    data: null,
                    loading: false,
                    error: 'Not load information'
                })
            }); 
    }, [url]);

    return state;
}
