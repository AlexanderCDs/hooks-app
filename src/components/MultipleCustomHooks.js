/** 
 * @author: alexandercddev
 * @description: 
 * @date: 30/Septiembre/2021
**/   
import React from 'react'
import { useFetch } from '../hooks/useFetch';
import { useCounter } from '../hooks/useCounter';

export const MultipleCustomHooks = () => {
    const { state: counter, increment, decrement } = useCounter(1);
    const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/characters/${counter}`); 
    const { name, occupation, img, status, nickname } = !!data && data[0]; // !!null == false, !null == true
    
    return (
        <div className="animate__animated animate__backInDown">
            <h1>Breaking Bad API</h1>
            <div className="card" style={{width: '100%'}}>
                {loading ? (   
                    <div className="alert alert-info text-center">
                        Loading...
                    </div>
                ) : (

                    <>
                        <img src={img} className="card-img-top" alt="..." />
                        <div className="card-body" style={{color: '#000'}}>
                            <div className="card-title" >
                                <h4>
                                    {name} 
                                </h4>
                               
                                
                            </div> 
                            <p className="card-text">
                                {`Nickname: ${nickname}`}
                            </p> 
                            <p className="card-text">
                                Status: <span className={`badge rounded-pill bg-${status === 'Alive' ? 'primary' : 'danger'}`}>{ status }</span>
                            </p> 
                        </div>
                        <ul className="list-group list-group-flush">
                            <li className="list-group-item">Occupation</li>
                            {occupation.map( (value, index) => (
                                <li key={'item-'+index} className="list-group-item">{ value }</li>
                            ))} 
                        </ul>
                    </>
                )} 
                <div className="stats">
                    {counter > 1 && (
                        <button 
                            className="btn btn-primary m-2" 
                            onClick={decrement}
                        >
                            Prev
                        </button>  
                    )}
                    
                    <button 
                        className="btn btn-primary m-2" 
                        onClick={increment}
                    >
                        Next
                    </button>  
                </div>
            </div>
        </div>
    )
}
