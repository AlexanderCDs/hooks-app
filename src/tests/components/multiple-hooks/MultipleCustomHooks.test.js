import React from 'react';
import {shallow} from 'enzyme';
import MultipleCustomHooks from '../../../components/MultipleCustomHooks'
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Test in MultipleCustomHooks', () => {
    useCounter.mockReturnValue({
        counter: 10, 
        increment: () => {},
        decrement: () => {}
    });

    test('should show sucess', () => {
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });
        const wrapper = shallow(<MultipleCustomHooks/>);
        expect(wrapper).toMatchSnapshot();
    });

    test('should show info', () => {
        useFetch.mockReturnValue({
            data: [{
                name: 'Alexander', 
                occupation: ['dev'], 
                img: '#', 
                status: 'Alive', 
                nickname: 'alexandercddev'
            }],
            loading: false,
            error: null
        });
        const wrapper = shallow(<MultipleCustomHooks/>);
        expect(wrapper.find('.alert').exists()).toBe(false);  
        expect(wrapper.find('.nickname').text().trim()).toBe('Nickname: alexandercddev');
        expect(wrapper.find('h4').text().trim()).toBe('Alexander');
        console.log(wrapper.html())
    });
})