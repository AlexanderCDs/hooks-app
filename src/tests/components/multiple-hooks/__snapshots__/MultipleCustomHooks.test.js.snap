// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in MultipleCustomHooks should show sucess 1`] = `
<div
  className="animate__animated animate__backInDown"
>
  <h1>
    Breaking Bad API
  </h1>
  <div
    className="card"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="alert alert-info text-center"
    >
      Loading...
    </div>
    <div
      className="stats"
    >
      <button
        className="btn btn-primary m-2"
        onClick={[Function]}
      >
        Next
      </button>
    </div>
  </div>
</div>
`;
